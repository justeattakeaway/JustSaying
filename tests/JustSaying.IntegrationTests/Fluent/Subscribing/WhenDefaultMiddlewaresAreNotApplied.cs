using JustSaying.Messaging.Middleware;
using JustSaying.TestingFramework;
using Microsoft.Extensions.DependencyInjection;
using Newtonsoft.Json;

namespace JustSaying.IntegrationTests.Fluent.Subscribing;

public class WhenDefaultMiddlewaresAreNotApplied(ITestOutputHelper outputHelper) : IntegrationTestBase(outputHelper)
{
    [AwsFact]
    public async Task Then_The_Pipeline_Should_Only_Contain_User_Specified_Middlewares()
    {
        // Arrange
        var handler = new InspectableHandler<SimpleMessage>();
        var testMiddleware = new InspectableMiddleware<SimpleMessage>();

        var services = GivenJustSaying()
            .ConfigureJustSaying((builder) =>
                MessagingBusBuilderTestExtensions.WithLoopbackTopic<SimpleMessage>(builder,
                    UniqueName,
                    c => c.WithMiddlewareConfiguration(m =>
                    {
                        m.Use(testMiddleware);
                        m.UseHandler(_ => handler);
                    })))
            .AddJustSayingHandlers(new[] { handler });

        string json = "";
        await WhenAsync(
            services,
            (_, listener, _, _) =>
            {
                dynamic middlewares = ((dynamic)listener.Interrogate().Data).Middleware;

                json = JsonConvert.SerializeObject(middlewares, Formatting.Indented)
                    .Replace(UniqueName, "TestQueueName");

                return Task.CompletedTask;
            });

        json.ShouldMatchApproved(c => c.SubFolder("Approvals"));
    }
}