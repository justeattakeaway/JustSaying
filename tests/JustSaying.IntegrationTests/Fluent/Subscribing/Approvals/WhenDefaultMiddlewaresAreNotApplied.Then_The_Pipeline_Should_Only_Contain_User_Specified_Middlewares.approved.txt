{
  "Middlewares": [
    {
      "MessageType": "SimpleMessage",
      "QueueName": "TestQueueName",
      "MiddlewareChain": [
        "InspectableMiddleware`1[SimpleMessage]",
        "HandlerInvocationMiddleware`1[JustSaying.TestingFramework.SimpleMessage]"
      ]
    }
  ]
}